FindMKL.cmake

```diff
@@ -0,0 +1,227 @@
+#[=======================================================================[.rst:
+FindMKL
+-------
+
+Example usage
+^^^^^^^^^^^^^
+
+  find_package(MKL REQUIRED OPTIONAL_COMPONENTS SCALAPACK)
+  target_link_libraries(... MKL::MKL)
+  target_link_libraries(... MKL::SCALAPACK)
+
+Note: Targets linking to MKL::SCALAPACK need not link to MKL::MKL. It is NOT 
+      an error if both targets are linked though.
+
+
+Options
+^^^^^^^
+
+``MKLROOT``
+  Environment variable set to MKL's root directory  
+
+``MKL_ROOT``
+  CMake variable set to MKL's root directory  
+
+``MKL_PARALLEL``           
+  ON|OFF   (default: ON / parallel)
+
+``MKL_64BIT``
+  ON|OFF   (default: OFF / 32bit interface)
+
+
+Components
+^^^^^^^^^^
+
+``SCALAPACK``
+
+
+Imported targets 
+^^^^^^^^^^^^^^^^
+
+``MKL::MKL``
+
+``MKL::SCALAPACK``
+
+
+Result variables
+^^^^^^^^^^^^^^^^
+
+``MKL_FOUND``
+  Found MKL.
+
+``MKL_SCALAPACK_FOUND``
+  Found ScaLAPACK.
+
+Not supported
+^^^^^^^^^^^^^
+
+- TBB threading back-end
+- F95 interfaces
+
+Note: Mixing GCC and Intel OpenMP backends is a bad idea.
+
+#]=======================================================================]
+
+# Modules
+#
+include(FindPackageHandleStandardArgs)
+
+# Functions
+#
+function(__mkl_find_library _name)
+    find_library(${_name}
+        NAMES ${ARGN}
+        HINTS ${MKL_ROOT}
+        PATH_SUFFIXES ${_mkl_libpath_suffix}
+                      lib
+        )
+    mark_as_advanced(${_name})
+endfunction()
+
+# Options
+#
+# The `NOT DEFINED` guards on CACHED variables are needed to make sure that 
+# normal variables with of same name always take precedence*.
```

Comment:
This is an interesting topic.

As far as I know, `if` does not check just normal variables, but also CACHE variables. So, once a cache variable with the same name is set, the normal one will not have priority over the cached one.

This has some minor edge cases that leads to problems in case you are working on the CMake side, however, I think that it may be a candidate solution for the problem in https://github.com/eth-cscs/cmake-recipes/pull/2#issuecomment-537828201

@halfflat @havogt what do you think about this solution?

Reply:
To be honest, I'm not sure I understand the desired semantics. IIRC, this code will:
* If the variable is in cache, don't do anything (e.g. from a previous run, or from a user-supplied `-Dvar=...`).
* If the variable is defined but not in cache, don't change, but also don't make a cache variable — this would be where at the calling site, something in the CMakeLists.txt has set it.
* If the variable is not defined at all, then make a cache variable with a reasonable default.

Is this correct?

Reply:
I think that your summary is complete and correct, and you are pointing out a few downsides of this solution that the one in #2 seems to solve.

I want to be able to set values from the script, so that in my package config script, when I do the `find_dependency`, I can force the FindMKL package to replicate exactly the same options as configured at build time. With this solution, values set from the script will be not considered if the user sets a cache variable with `-D` from the command line (as you are pointing out in your first point).

Moreover, with the alternative solution in #2 of using "*_CUSTOM" variables as prioritized, related CACHE variables are always set to show to the user the final value that will be used. Anyway, this may still be a bit misleading because it discards any change made by the user to the CACHE variables.

Reply:
@albestro I don't see the downsides here. Can you explain your exact use-case where this is a problem?

Reply:
@albestro , the behaviour that @havogt outlined is exactly the one I wanted. I don't see a downside tbh

> To be honest, I'm not sure I understand the desired semantics. IIRC, this code will:
> 
>     * If the variable is in cache, don't do anything (e.g. from a previous run, or from a user-supplied `-Dvar=...`).
> 
>     * If the variable is defined but not in cache, don't change, but also don't make a cache variable — this would be where at the calling site, something in the CMakeLists.txt has set it.
> 
>     * If the variable is not defined at all, then make a cache variable with a reasonable default.
> 
> 
> Is this correct?

@havogt That is correct. That is exactly the behaviour that I was going for.

> I want to be able to set values from the script, so that in my package config script, when I do the find_dependency, I can force the FindMKL package to replicate exactly the same options as configured at build time. With this solution, values set from the script will be not considered if the user sets a cache variable with -D from the command line (as you are pointing out in your first point).

@albestro My understanding is that values from the script will take precedence over `-D` options since they are normal CMake variables as opposed to CACHE variables. In other words, yes, the values from the script will be considered.

Reply:
Yes, I was wrong. I got confused, I was not considering the right order of referencing a variable name: the order is `normal variable > cache variable`, so if there is any normal variable, it will always have priority no matter if a cache variable exists or not. And it is exactly what I want to achieve in my use-case.

About not exposing the value in the cache if a value is set from the script, it can be ok: it is true that the user cannot see what the final value is, but it is also true that he/she does not get confused by a cache variable that gets always set to the value decided in the script.
In order to have this last behavior with CMake version <3.15, where they added the [CACHE{variable}](https://cmake.org/cmake/help/v3.15/manual/cmake-language.7.html#variable-references), I can just think of workarounds that are not worth it.

I'm very satisfied with this work. Thanks everyone! I'm going to employ it in my script.